### PR TITLE
[DataGrid] First iteration on filtering, basic support

### PR DESCRIPTION
**Pull request by dtassone**
Work on this problem: #201.

TODO

- [x] toggle off by default
- [x] Filters for numbers, dates and dateTime
- [ ] components customization

### CODE REVIEW

File: lerna.json [hunk 1]
```diff
@@ -5,11 +5,7 @@
       "yes": true
     }
   },
-  "packages": [
-    "!packages/grid",
-    "packages/*",
-    "packages/grid/*"
-  ],
+  "packages": ["!packages/grid", "packages/*", "packages/grid/*"],
```
Comment (oliviertassinari) [inline lerna.json:8]: Unrelated change

File: packages/grid/_modules_/grid/components/column-header-item.tsx [hunk 2]
```diff
@@ -100,26 +111,35 @@ export const ColumnHeaderItem = ({
...
-            columnWidth={width}
-          />
-        )}
-        {column.type !== 'number' && (
-          <ColumnHeaderSortIcon
-            direction={sortDirection}
-            index={sortIndex}
-            hide={column.hideSortIcons}
-          />
+
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/column-header-item.tsx:117]:
```suggestion
```

File: packages/grid/_modules_/grid/components/column-header-item.tsx [hunk 3]
```diff
@@ -100,26 +111,35 @@ export const ColumnHeaderItem = ({
...
+        <div className={'MuiDataGrid-colCellTitleContainer'}>
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/column-header-item.tsx:118]:
```suggestion
        <div className="MuiDataGrid-colCellTitleContainer">
```

File: packages/grid/_modules_/grid/components/icons/index.tsx [hunk 4]
```diff
@@ -11,4 +11,46 @@ export const ArrowDownwardIcon = createSvgIcon(
   'ArrowDownward',
 );
 
+export const FilterIcon = createSvgIcon(
+  <path d="M10 18h4v-2h-4v2zM3 6v2h18V6H3zm3 7h12v-2H6v2z" />,
+  'FilterList',
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/icons/index.tsx:16]: To match `FilterIcon`'s variable

```suggestion
  'Filter',
```

File: packages/grid/_modules_/grid/components/icons/index.tsx [hunk 5]
```diff
@@ -11,4 +11,46 @@ export const ArrowDownwardIcon = createSvgIcon(
...
+export const CloseIcon = createSvgIcon(
+  <path d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z" />,
+  'Close',
+);
+
+export const AddIcon = createSvgIcon(<path d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z" />, 'Add');
+
+export const LoadIcon = createSvgIcon(
+  <path d="M12 4V1L8 5l4 4V6c3.31 0 6 2.69 6 6 0 1.01-.25 1.97-.7 2.8l1.46 1.46C19.54 15.03 20 13.57 20 12c0-4.42-3.58-8-8-8zm0 14c-3.31 0-6-2.69-6-6 0-1.01.25-1.97.7-2.8L5.24 7.74C4.46 8.97 4 10.43 4 12c0 4.42 3.58 8 8 8v3l4-4-4-4v3z" />,
+  'Loop',
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/icons/index.tsx:55]:
```suggestion
  'Load',
```

File: packages/grid/_modules_/grid/components/icons/index.tsx [hunk 6]
```diff
@@ -11,4 +11,46 @@ export const ArrowDownwardIcon = createSvgIcon(
...
+);
+
+export const ViewWeekIcon = createSvgIcon(
+  <path d="M6 5H3c-.55 0-1 .45-1 1v12c0 .55.45 1 1 1h3c.55 0 1-.45 1-1V6c0-.55-.45-1-1-1zm14 0h-3c-.55 0-1 .45-1 1v12c0 .55.45 1 1 1h3c.55 0 1-.45 1-1V6c0-.55-.45-1-1-1zm-7 0h-3c-.55 0-1 .45-1 1v12c0 .55.45 1 1 1h3c.55 0 1-.45 1-1V6c0-.55-.45-1-1-1z" />,
+  'ViewWeek',
+);
+
 export const SeparatorIcon = createSvgIcon(<path d="M11 19V5h2v14z" />, 'Separator');
+
+export const TripleDotsVertical = createSvgIcon(
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/icons/index.tsx:41]:
```suggestion
export const TripleDotsVerticalIcon = createSvgIcon(
```

File: packages/grid/_modules_/grid/components/icons/index.tsx [hunk 7]
```diff
@@ -11,4 +11,46 @@ export const ArrowDownwardIcon = createSvgIcon(
...
+  <path d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z" />,
+  'MoreVert',
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/icons/index.tsx:43]:
```suggestion
  'TripleDotsVertical',
```

File: packages/grid/_modules_/grid/components/menu/columnMenu/GridColumnHeaderMenu.tsx [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+  return (
+    <GridMenu
+      open={gridState.columnMenu.open}
+      target={target}
+      onKeyDown={handleListKeyDown}
+      onClickAway={hideMenuDelayed}
+    >
+      <SortMenuItems onClick={hideMenu} column={currentColumn!} />
+      <FilterMenuItem onClick={hideMenu} column={currentColumn!} />
+
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/menu/columnMenu/GridColumnHeaderMenu.tsx:87]:
```suggestion
```

File: packages/grid/_modules_/grid/components/menu/columnMenu/SortMenuItems.tsx [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+  const sortDirection = React.useMemo(() => {
+    if (!column) {
+      return null;
+    }
+    const sortItem = sortModel.find((item) => item.field === column.field);
+    return sortItem?.sort;
+  }, [column, sortModel]);
+
+  const onSortMenuItemClick = React.useCallback(
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/menu/columnMenu/SortMenuItems.tsx:26]: IMHO, we shouldn't have a usecallback here

https://kentcdodds.com/blog/usememo-and-usecallback

File: packages/grid/_modules_/grid/components/menu/columnMenu/SortMenuItems.tsx [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+  if (!column || !column.sortable) {
+    return null;
+  }
+
+  return (
+    <React.Fragment>
+      <MenuItem onClick={onSortMenuItemClick} disabled={sortDirection == null}>
+        Unsort
+      </MenuItem>
+      <MenuItem onClick={onSortMenuItemClick} data-value={'asc'} disabled={sortDirection === 'asc'}>
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/menu/columnMenu/SortMenuItems.tsx:44]: This would be simpler and as performant

```suggestion
      <MenuItem onClick={onSortMenuItemClick('asc')} disabled={sortDirection === 'asc'}>
```

File: packages/grid/_modules_/grid/components/menu/columnMenu/SortMenuItems.tsx [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+        Sort By Asc
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/menu/columnMenu/SortMenuItems.tsx:45]: I would suggest we start to anticipate for https://material-ui.com/components/data-grid/localization/#locale-text with: 
```suggestion
        {t('Sort By Asc')}
```

t would be an identity function, one that we can grep in the future, easier to find all the places where translated text is used.

File: packages/grid/_modules_/grid/components/tools/FilterForm.tsx [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+import { filterableColumnsSelector } from '../../hooks/features/columns/columnsSelector';
+import { useGridSelector } from '../../hooks/features/core/useGridSelector';
+import { FilterItem, LinkOperator } from '../../hooks/features/filter/visibleRowsState';
+import { ColDef } from '../../models/colDef/colDef';
+import { ApiContext } from '../api-context';
+import { CloseIcon, LoadIcon } from '../icons/index';
+
+export interface FilterFormProps {
+  item: FilterItem;
+
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/tools/FilterForm.tsx:19]:
```suggestion
```

File: packages/grid/_modules_/grid/components/tools/FilterForm.tsx [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+  showMultiFilterOperators?: boolean;
+  multiFilterOperator?: LinkOperator;
+  disableMultiFilterOperator?: boolean;
+
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/tools/FilterForm.tsx:23]:
```suggestion
```

File: packages/grid/_modules_/grid/components/tools/FilterForm.tsx [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+    },
+    [applyMultiFilterOperatorChanges],
+  );
+
+  const handleDeleteFilter = React.useCallback(() => {
+    deleteFilter(item);
+  }, [deleteFilter, item]);
+
+  return (
+    <div style={{ display: 'flex', justifyContent: 'space-around', padding: '10px' }}>
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/tools/FilterForm.tsx:89]: inline style

File: packages/grid/_modules_/grid/components/tools/Preferences.tsx [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+      placement="bottom"
+      open={gridState.preferencePanel.open}
+      anchorEl={target || apiRef?.current.rootElementRef!.current}
+      style={{ position: 'relative' }}
+    >
+      <div style={{ position: 'absolute', left: 0, top: 0, zIndex: 10000 }}>
+        <IconButton color="primary" aria-label="Close" component="span" onClick={hidePreferences}>
+          <CloseIcon />
+        </IconButton>
+      </div>
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/tools/Preferences.tsx:129]: This is my number one UX issue so far playing with the filter. It took me some time to figure out how to hide this *** popup 😆 . And even now that I know how to do it, oh god I hate it 😇.
I think that we need 1. the `ClickAwayListener`, 2. handle the <kbd>Escape</kbd> key. I would propose we use a native select, for now (`<Select native />`).

File: packages/grid/_modules_/grid/models/api/sortApi.ts [hunk 16]
```diff
@@ -23,4 +24,6 @@ export interface SortApi {
    * @param handler
    */
   onSortModelChange: (handler: (param: SortModelParams) => void) => () => void;
+
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/models/api/sortApi.ts:27]: Miss description

File: packages/grid/_modules_/grid/models/gridOptions.tsx [hunk 17]
```diff
@@ -17,6 +23,14 @@ import { SortDirection, SortModel } from './sortModel';
  * Set of icons used in the grid component UI.
  */
 export interface IconsOptions {
+  /**
+   * Icon displayed on the side of the column header title to display the filter input component.
+   */
+  columnMenu?: React.ElementType;
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/models/gridOptions.tsx:29]: API convention, components are uppercased.

```suggestion
  ColumnMenu?: React.ElementType;
```

File: packages/grid/_modules_/grid/models/gridOptions.tsx [hunk 18]
```diff
@@ -17,6 +23,14 @@ import { SortDirection, SortModel } from './sortModel';
...
+  /**
+   * Icon displayed on the column menu filter tab.
+   */
+  columnFiltering?: React.ElementType;
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/models/gridOptions.tsx:33]:
```suggestion
  ColumnFiltering?: React.ElementType;
```

**Review by oliviertassinari — CHANGES_REQUESTED** [review]
A few quick feedback :)

File: packages/grid/_modules_/grid/components/tools/Preferences.tsx (hunk 15 above)
Comment (dtassone) [reply packages/grid/_modules_/grid/components/tools/Preferences.tsx:129]: Can we have a proper fix instead of using the native select?

File: lerna.json (hunk 1 above)
Comment (dtassone) [reply lerna.json:8]: prettier...

File: packages/grid/_modules_/grid/components/tools/Preferences.tsx (hunk 15 above)
Comment (oliviertassinari) [reply packages/grid/_modules_/grid/components/tools/Preferences.tsx:129]: I'm aware of two solutions:

1. native slider: https://material-ui.com/components/selects/#native-select, the UI is still OK
2. #441

I can't think of anything else.

Comment (dtassone) [reply packages/grid/_modules_/grid/components/tools/Preferences.tsx]: using Native for now

File: packages/grid/_modules_/grid/components/column-header-menu-icon.tsx [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+    const apiRef = React.useContext(ApiContext);
+    const [, setGridState, forceUpdate] = useGridState(apiRef!);
+
+    const menuIconClick = React.useCallback(() => {
+      setGridState((state) => ({
+        ...state,
+        columnMenu: { open: true, field: column.field },
+        preferencePanel: { open: false },
+      }));
+      forceUpdate();
```
Comment (DanailH) [inline packages/grid/_modules_/grid/components/column-header-menu-icon.tsx:25]: I don't remember if we had a duscussion about having the `forceUpdate` being called from within the `setGridState`. It seems that we need to call it every time the `setGridState` is called.

File: packages/grid/_modules_/grid/components/menu/columnMenu/GridColumnHeaderMenu.tsx [hunk 20]
```diff
@@ -0,0 +1,96 @@
...
+    <GridMenu
+      open={gridState.columnMenu.open}
+      target={target}
+      onKeyDown={handleListKeyDown}
+      onClickAway={hideMenuDelayed}
+    >
+      <SortMenuItems onClick={hideMenu} column={currentColumn!} />
+      <FilterMenuItem onClick={hideMenu} column={currentColumn!} />
+      <MenuItem onClick={hideMenuDelayed} disabled>
+        Auto size
```
Comment (DanailH) [inline packages/grid/_modules_/grid/components/menu/columnMenu/GridColumnHeaderMenu.tsx:88]: Are we doing the translation handler as @oliviertassinari suggested in one of the previous comments?

File: packages/grid/_modules_/grid/components/tools/FilterForm.tsx [hunk 21]
```diff
@@ -0,0 +1,178 @@
...
+          ))}
+        </Select>
+      </FormControl>
+      <FormControl className={classes.FilterValueInput}>
+        {item.operator &&
+          React.createElement(item.operator?.InputComponent, {
+            item,
+            applyValue: applyFilterChanges,
+          })}
+      </FormControl>
```
Comment (DanailH) [inline packages/grid/_modules_/grid/components/tools/FilterForm.tsx:165]:
```suggestion
        {item.operator && (
          <FormControl className={classes.FilterValueInput}>
            {React.createElement(item.operator?.InputComponent, {
            item,
            applyValue: applyFilterChanges,
          })}
        </FormControl>
      )}
```

This way there won't be an empty DOM element if `item.operator` is not provided.

File: packages/grid/_modules_/grid/hooks/features/filter/useFilter.ts [hunk 22]
```diff
@@ -0,0 +1,202 @@
...
+  LinkOperator,
+} from './visibleRowsState';
+
+export const useFilter = (apiRef: ApiRef): void => {
+  const logger = useLogger('useFilter');
+  const [gridState, setGridState, forceUpdate] = useGridState(apiRef);
+
+  const rows = useGridSelector(apiRef, sortedRowsSelector);
+  const filterableColumns = useGridSelector(apiRef, filterableColumnsSelector);
+  // const visibleRowsState = useGridSelector(apiRef, visibleRowsStateSelector);
```
Comment (DanailH) [inline packages/grid/_modules_/grid/hooks/features/filter/useFilter.ts:25]:
```suggestion
```

File: packages/grid/_modules_/grid/components/menu/columnMenu/GridColumnHeaderMenu.tsx [hunk 23]
```diff
@@ -0,0 +1,96 @@
...
+
+  // TODO: Fix issue with portal in V5
+  const hideTimeout = React.useRef<any>();
+  const hideMenu = React.useCallback(() => {
+    setGridState((state) => ({ ...state, columnMenu: { open: false } }));
+    forceUpdate();
+  }, [forceUpdate, setGridState]);
+
+  const hideMenuDelayed = React.useCallback(() => {
+    hideTimeout.current = setTimeout(() => hideMenu(), 50);
```
Comment (DanailH) [inline packages/grid/_modules_/grid/components/menu/columnMenu/GridColumnHeaderMenu.tsx:42]:
```suggestion
    hideTimeout.current = setTimeout(hideMenu, 50);
```

File: packages/grid/_modules_/grid/components/menu/columnMenu/GridColumnHeaderMenu.tsx (hunk 20 above)
Comment (dtassone) [reply packages/grid/_modules_/grid/components/menu/columnMenu/GridColumnHeaderMenu.tsx:88]: I would prefer to do it in a separate PR

File: packages/grid/_modules_/grid/components/column-header-menu-icon.tsx (hunk 19 above)
Comment (dtassone) [reply packages/grid/_modules_/grid/components/column-header-menu-icon.tsx:25]: Not in every case, especially in useRows

File: packages/grid/_modules_/grid/components/menu/columnMenu/GridColumnHeaderMenu.tsx (hunk 20 above)
Comment (oliviertassinari) [reply packages/grid/_modules_/grid/components/menu/columnMenu/GridColumnHeaderMenu.tsx:88]: To be clear, the idea of the handler was meant at a signal, we could use something else, like a comment.

File: packages/grid/_modules_/grid/components/styled-wrappers/GridRootStyles.ts [hunk 24]
```diff
@@ -147,6 +156,13 @@ export const useStyles = makeStyles(
         '& .MuiDataGrid-iconSeparator': {
           color: 'inherit',
         },
+        '& .MuiDataGrid-menuIcon': {
+          visibility: 'hidden',
+          fontSize: 18,
+        },
+        '& .MuiDataGrid-colCell:hover  .MuiDataGrid-menuIcon': {
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/styled-wrappers/GridRootStyles.ts:163]: Two spaces

File: packages/grid/_modules_/grid/components/styled-wrappers/GridRootStyles.ts [hunk 25]
```diff
@@ -100,16 +99,26 @@ export const useStyles = makeStyles(
...
         },
         '& .MuiDataGrid-sortIcon': {
           fontSize: 18,
         },
-        '& .MuiDataGrid-colCellCenter': {
+        '& .MuiDataGrid-colCellCenter .MuiDataGrid-colCellTitleContainer': {
           justifyContent: 'center',
         },
-        '& .MuiDataGrid-colCellRight': {
+        '& .MuiDataGrid-colCellRight .MuiDataGrid-colCellTitleContainer': {
```
[the remark this answers is not in the document]
Comment (oliviertassinari) [reply packages/grid/_modules_/grid/components/styled-wrappers/GridRootStyles.ts:121]: If there an opportunity to flatten the CSS specificity or reduce the amount of DOM nodes we should take it.

**Comment by dtassone** [comment]
Todo
- components customization
- Filters for numbers, dates and dateTime

File: packages/grid/_modules_/grid/components/tools/FilterPanel.tsx [hunk 26]
```diff
@@ -0,0 +1,101 @@
...
+            deleteFilter={deleteFilter}
+            showMultiFilterOperators={hasMultipleFilters && index > 0}
+            multiFilterOperator={gridState.filter.linkOperator}
+            disableMultiFilterOperator={index !== 1}
+            applyMultiFilterOperatorChanges={applyFilterLinkOperator}
+          />
+        ))}
+      </div>
+      <div className={classes.filterPanelFooter}>
+        <Button onClick={addNewFilter} startIcon={<AddIcon />}>
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/tools/FilterPanel.tsx:92]: in v5, we have removed the default "default" color. So, in order to anticipate the change:

```suggestion
        <Button onClick={addNewFilter} startIcon={<AddIcon />} color="primary">
```

x2

File: packages/grid/_modules_/grid/components/tools/Preferences.tsx [hunk 27]
```diff
@@ -0,0 +1,141 @@
...
+    <Popper
+      placement="bottom"
+      open={preferencePanelState.open}
+      anchorEl={target || apiRef?.current.rootElementRef!.current}
+      style={{ position: 'relative' }}
+    >
+      <ClickAwayListener onClickAway={hidePreferences}>
+        <Paper
+          square
+          className={classes.root}
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/tools/Preferences.tsx:99]: This is not the root
```suggestion
          className={classes.paper}
```

File: packages/grid/data-grid/src/DataGrid.test.tsx [hunk 28]
```diff
@@ -25,7 +25,7 @@ describe('<DataGrid />', () => {
         brand: 'Puma',
       },
     ],
-    columns: [{ field: 'brand', width: 100 }],
+    columns: [{ field: 'brand', width: 180 }],
```
Comment (oliviertassinari) [inline packages/grid/data-grid/src/DataGrid.test.tsx:28]: Would this be simpler? 100 is the default for all the types.

```suggestion
    columns: [{ field: 'brand' }],
```

File: packages/grid/data-grid/src/DataGrid.test.tsx [hunk 29]
```diff
@@ -274,7 +274,9 @@ describe('<DataGrid />', () => {
             <DataGrid {...defaultProps} />
           </div>,
         );
-        const header = screen.getByRole('columnheader', { name: 'brand' });
+        const header = screen
+          .getByRole('columnheader', { name: 'brand' })
+          .querySelector('.MuiDataGrid-colCellTitleContainer');
```
Comment (oliviertassinari) [inline packages/grid/data-grid/src/DataGrid.test.tsx:279]: Does it mean we have broken the a11y feature?

File: packages/grid/_modules_/grid/models/api/sortApi.ts [hunk 30]
```diff
@@ -23,4 +24,10 @@ export interface SortApi {
    * @param handler
    */
   onSortModelChange: (handler: (param: SortModelParams) => void) => () => void;
+  /**
+   * Set the sort direction of a column
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/models/api/sortApi.ts:28]:
```suggestion
   * Set the sort direction of a column.
```

File: packages/grid/_modules_/grid/models/api/preferencesPanelApi.ts [hunk 31]
```diff
@@ -0,0 +1,16 @@
+import { PreferencePanelsValue } from '../../hooks/features/preferencesPanel/preferencesPanelValue';
+
+/**
+ * The PreferencesPanel API interface that is available in the grid [[apiRef]].
+ */
+export interface PreferencesPanelApi {
+  /**
+   * Display the preferences panel with the PreferencePanelsValue opened .
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/models/api/preferencesPanelApi.ts:8]:
```suggestion
   * Display the preferences panel with the PreferencePanelsValue opened.
```

File: packages/grid/_modules_/grid/components/tools/Preferences.tsx [hunk 32]
```diff
@@ -0,0 +1,141 @@
...
+    width: 600,
+    minHeight: 200,
+    display: 'flex',
+    flexDirection: 'column',
+    '& .tab': {
+      minWidth: 50,
+    },
+    '& .chip': {
+      margin: 2,
+    },
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/tools/Preferences.tsx:41]: It seems to be dead code.

```suggestion
```

File: packages/grid/_modules_/grid/components/tools/Preferences.tsx [hunk 33]
```diff
@@ -0,0 +1,141 @@
...
+import { ViewWeekIcon } from '../icons/index';
+import { FilterPanel } from './FilterPanel';
+
+export function TabPanel(props) {
+  const { children } = props;
+
+  return (
+    <div
+      role="tabpanel"
+      style={{ display: 'flex', flexDirection: 'column', overflow: 'hidden', flex: 1 }}
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/tools/Preferences.tsx:23]: I'm not a fan 😁

File: packages/grid/_modules_/grid/components/tools/Preferences.tsx [hunk 34]
```diff
@@ -0,0 +1,141 @@
+import { ClickAwayListener, Paper, Popper, Tab, Tabs, Theme } from '@material-ui/core';
+import { makeStyles } from '@material-ui/core/styles';
+import * as React from 'react';
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/tools/Preferences.tsx:3]: I think that we should have nested imports to minimize size in dev mode.

```suggestion
import * as React from 'react';
import { makeStyles, Theme } from '@material-ui/core/styles';
import ClickAwayListener from '@material-ui/core/ClickAwayListener';
import Paper from '@material-ui/core/Paper';
import Popper from '@material-ui/core/Popper';
import Tab from '@material-ui/core/Tab';
import Tabs from '@material-ui/core/Tabs';
```

File: packages/grid/_modules_/grid/components/tools/FilterPanel.tsx [hunk 35]
```diff
@@ -0,0 +1,101 @@
...
+          Filter
+        </Button>
+        <Button onClick={clearFilter} startIcon={<CloseIcon />}>
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/tools/FilterPanel.tsx:95]: We will need to see how to best align the buttons.

File: packages/grid/_modules_/grid/components/tools/FilterForm.tsx [hunk 36]
```diff
@@ -0,0 +1,190 @@
...
+      <FormControl className={classes.FilterValueInput}>
+        {currentOperator &&
+          React.createElement(currentOperator.InputComponent, {
+            item,
+            applyValue: applyFilterChanges,
+          })}
+      </FormControl>
+      <FormControl>
+        <IconButton
+          color="primary"
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/tools/FilterForm.tsx:180]: I think that we should reduce the attention we give to this button

```suggestion
```

File: packages/grid/_modules_/grid/components/tools/FilterForm.tsx [hunk 37]
```diff
@@ -0,0 +1,190 @@
...
+          aria-label="Delete"
+          component="span"
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/tools/FilterForm.tsx:182]: Why not render a `<button>`?
```suggestion
```

File: packages/grid/_modules_/grid/components/tools/FilterForm.tsx [hunk 38]
```diff
@@ -0,0 +1,190 @@
...
+        {currentOperator &&
+          React.createElement(currentOperator.InputComponent, {
+            item,
+            applyValue: applyFilterChanges,
+          })}
+      </FormControl>
+      <FormControl>
+        <IconButton
+          color="primary"
+          aria-label="Delete"
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/tools/FilterForm.tsx:181]: No need to capitalize, but we can use the same translation for the title:

```suggestion
          aria-label="Delete"
          title="Delete"
```

File: packages/grid/_modules_/grid/components/tools/FilterForm.tsx [hunk 39]
```diff
@@ -0,0 +1,190 @@
...
+          component="span"
+          onClick={handleDeleteFilter}
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/tools/FilterForm.tsx:183]: I would suggest we reduce its size:

```suggestion
          onClick={handleDeleteFilter}
          size="small"
```

File: packages/grid/_modules_/grid/components/tools/Preferences.tsx [hunk 40]
```diff
@@ -0,0 +1,141 @@
...
+      <ClickAwayListener onClickAway={hidePreferences}>
+        <Paper
+          square
+          className={classes.root}
+          style={{
+            maxHeight: viewportSizes.height,
+            maxWidth: viewportSizes.width,
+          }}
+        >
+          <Tabs
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/tools/Preferences.tsx:105]: Something is wrong with multiple items and a limited vertical space:

Capture d’écran 2020-11-18 à 20 25 54

I would add a `flex-shrink: 0;`

File: packages/grid/_modules_/grid/components/tools/StringFilterInputValue.tsx [hunk 41]
```diff
@@ -0,0 +1,44 @@
+import { TextField } from '@material-ui/core';
+import * as React from 'react';
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/components/tools/StringFilterInputValue.tsx:2]:
```suggestion
import * as React from 'react';
import TextField from '@material-ui/core/TextField';
```

**Review by oliviertassinari — CHANGES_REQUESTED** [review]
We are getting closer :D but I can't open the menu in https://deploy-preview-411--material-ui-x.netlify.app/components/data-grid/.

File: packages/grid/_modules_/grid/components/menu/GridMenu.tsx [hunk 42]
```diff
@@ -0,0 +1,36 @@
...
+
+export const GridMenu: React.FC<MenuProps> = ({
+  open,
+  target,
+  onKeyDown,
+  onClickAway,
+  children,
+}) => {
+  return (
+    <Popper open={open} anchorEl={target as any} transition>
```
[the remark this answers is not in the document]
Comment (dtassone) [reply packages/grid/_modules_/grid/components/menu/GridMenu.tsx:19]: what is x?

[the remark this answers is not in the document]
Comment (oliviertassinari) [reply packages/grid/_modules_/grid/components/menu/GridMenu.tsx:19]: The alignment of the column.

File: packages/grid/data-grid/src/DataGrid.test.tsx (hunk 29 above)
Comment (dtassone) [reply packages/grid/data-grid/src/DataGrid.test.tsx:279]: No the sorting is just on the title as we now have different buttons in the cell

**Comment by dtassone** [comment]
> We are getting closer :D but I can't open the menu in https://deploy-preview-411--material-ui-x.netlify.app/components/data-grid/.

Not sure why 🤔 
It's working locally

image

File: packages/grid/_modules_/grid/models/colDef/colDef.ts [hunk 43]
```diff
@@ -116,6 +117,19 @@ export interface ColDef {
...
+   * Allows to disable the column menu for this column.
+   */
+  disableColumnMenu?: boolean;
+  /**
+   * If `true`, the column is filterable.
+   * @default true
+   */
+  filterable?: boolean;
+  /**
+   * Allows to set the filterOperators for this column
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/models/colDef/colDef.ts:130]:
```suggestion
   * Allows setting the filter operators for this column.
```

**Review by oliviertassinari — APPROVED** [review]
I have noticed some issues like:

- Missing documentation
- Placement of menu that can be improved
- Jump when closing the menu jump
- Wrong elevation elevation
- Wrong position position

but I think that we could solve them in follow-up. Better keep pull requests small

**Review by DanailH — APPROVED** [review]
